woocommerce/woocommerce-gateway-stripe
# Fix address fields mapping for Google Pay and UAE addresses

Fixes #3430 

## Changes proposed in this Pull Request:
When using Google Pay with a UAE shipping address, we get the error `Shipping Town / City is a required field.` on checkout. This is because Google Pay stores the emirate in "region", [which gets mapped to the "state" field](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7108f35a2aaebce59914f5dacc3bfe01dab2c1a9/client/blocks/normalize.js#L75) in WooCommerce. For UAE, WooCommerce [expects the emirate to be stored in the "city" field](https://github.com/woocommerce/woocommerce/blob/d87c3d38d72c12469f1bf9c35364d77076a93e66/plugins/woocommerce/includes/class-wc-countries.php#L856).

This PR adds an address field mapping step on the checkout form data if the shipping or billing country is UAE. The approach is similar to [a corrective step that we do for Apple Pay and HK](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7108f35a2aaebce59914f5dacc3bfe01dab2c1a9/includes/payment-methods/class-wc-stripe-payment-request.php#L1492).







## Testing instructions
1. Add a UAE address in your Google Pay account: https://payments.google.com/gp/w/u/0/home/addressbook
   - For example: PO Box 12345, Abu Dhabi, United Arab Emirates
2. Make sure you are NOT in the Google Pay API Test Cards allowlist: https://groups.google.com/g/googlepay-test-mode-stub-data
   - If you are in the allowlist Google group, you will have access to the test address data set only, which does not include UAE.
3. Make sure test mode is enabled in Stripe, to avoid actual charges: `wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`
4. As a shopper, add a product to your cart, and go to checkout.
   - **Important**: Make sure the shipping and biling form fields are clear. If they are pre-populated (due to autofill), clear them out first. Reason: We have added behavior in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2886 to pull in data from the checkout form for required fields. If you have been testing with a different address that has a city, this may get pulled in.
7. Click the "Buy with Google Pay" button to open the modal.
8. Make sure to choose the UAE address as the shipping address.
9. Click Pay.
10. Before fix: Notice the payment fails with the error "Shipping Town / City is a required field."
11. Fixed: Your order should succeed. Verify that it shows the correct shipping address.






**Before: Checkout fails because the "city" field is empty.**
<img width="500" alt="Screenshot 2024-09-10 at 3 52 12 PM" src="https://github.com/user-attachments/assets/57e6ae09-ef9f-4595-981c-d06f8e2d7b2b">


---


**Post merge**
